core/src/services/dbfs/reader.rs

```diff
@@ -0,0 +1,311 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::cmp::Ordering;
+use std::io;
+use std::mem;
+use std::str::FromStr;
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+
+use base64::engine::general_purpose;
+use base64::Engine;
+use bytes::Buf;
+use bytes::BufMut;
+use bytes::Bytes;
+use futures::StreamExt;
+use futures::TryStreamExt;
+use http::Request;
+use http::Response;
+use serde::Deserialize;
+
+use crate::raw::*;
+use crate::*;
+
+impl HttpClient {
```

Comment:
Implementing API for HttpClient is discouraged. We will need to:

- Implement `dbfs_read` which accept `offset` and `length`
- DbfsReader should hold the `DbfsCore` (take s3 impelmentaion as an example)
- Implement `oio::Read` for `DbfsReader` directly

We don't need to play with `IncomingAsyncBody`.

Reply:
Hi @Xuanwo, this was my initial idea when implementing the Reader. However, I later discovered that in the operator `read_with`, the buffer length is generated from `RpRead`, which is invoked before `reader.read_exact()`. So, how can we determine the content length before calling `poll_read()` in DbfsReader? (As you can see, I ended up to calling the status API in the DbfsBackend read method to obtain the file length in advance.)

Reply:
> So, how can we determine the content length before calling `poll_read()` in DbfsReader?

We can get the content length before creating `DbfsReader`.

Reply:
I've roughly implemented the reader's `poll_read`. Since the buf sent from the operator will reset every time when `poll_read` is called, we store our response inside the DbfsReader's internal buffer. Once we have the complete length, we can write the data into the buffer. However, for `poll_seek`, I don't understand what you mean by "drop the existing buffer" ?

Reply:
> However, for `poll_seek`, I don't understand what you mean by "drop the existing buffer" ?

This PR is too large for now. How about implementing `poll_seek` in next PR?